c7n/filters/policystatement.py

```diff
@@ -0,0 +1,101 @@
+# Copyright The Cloud Custodian Authors.
+# SPDX-License-Identifier: Apache-2.0
+import json
+
+from .core import Filter
+from c7n.utils import type_schema, format_string_values
+
+
+class HasStatementFilter(Filter):
+    """Find resources with matching access policy statements.
+    :example:
+    .. code-block:: yaml
+            policies:
+              - name: sns-check-statement-id
+                resource: sns
+                filters:
+                  - type: has-statement
+                    statement_ids:
+                      - BlockNonSSL
+            policies:
+              - name: sns-check-block-non-ssl
+                resource: sns
+                filters:
+                  - type: has-statement
+                    statements:
+                      - Effect: Deny
+                        Action: 'SNS:Publish'
+                        Principal: '*'
+                        Condition:
+                            Bool:
+                                "aws:SecureTransport": "false"
+    """
+    schema = type_schema(
+        'has-statement',
+        statement_ids={'type': 'array', 'items': {'type': 'string'}},
+        statements={
+            'type': 'array',
+            'items': {
+                'type': 'object',
+                'properties': {
+                    'Sid': {'type': 'string'},
+                    'Effect': {'type': 'string', 'enum': ['Allow', 'Deny']},
+                    'Principal': {'anyOf': [
+                        {'type': 'string'},
+                        {'type': 'object'}, {'type': 'array'}]},
+                    'NotPrincipal': {
+                        'anyOf': [{'type': 'object'}, {'type': 'array'}]},
+                    'Action': {
+                        'anyOf': [{'type': 'string'}, {'type': 'array'}]},
+                    'NotAction': {
+                        'anyOf': [{'type': 'string'}, {'type': 'array'}]},
+                    'Resource': {
+                        'anyOf': [{'type': 'string'}, {'type': 'array'}]},
+                    'NotResource': {
+                        'anyOf': [{'type': 'string'}, {'type': 'array'}]},
+                    'Condition': {'type': 'object'}
+                },
+                'required': ['Effect']
+            }
+        })
+
+    def process(self, resources, event=None):
+        return list(filter(None, map(self.process_resource, resources)))
+
+    def process_resource(self, resource):
+        p = resource.get('Policy')
```

Comment:
While `Policy` is a common attribute, IAM policies can show up in different spots across resources. We may be able to leverage `policy_attribute` and `get_resource_policy` to allow resources to override where/how their IAM policies are defined. The example [here](https://github.com/cloud-custodian/cloud-custodian/pull/6003/files#diff-30de4ac0839408f56b1e337821788581c2b1ecf95d09bcdac4f2dc59c5081956R440-R441) might be useful.